src/library/vm/vm.h

```diff
@@ -152,6 +158,9 @@ class vm_external : public vm_obj_cell {
     virtual ~vm_external() {}
     virtual vm_external * ts_clone(vm_clone_fn const &) = 0;
     virtual vm_external * clone(vm_clone_fn const &) = 0;
+    /** A hash function for externals so that equality of vm objects can be checked.
+     * [hack] this should be made abstract but there are too many inheritors. */
```

Comment:
There are 35 subclasses of `vm_external` in total.  Just copy&paste this line everywhere.  It's probably better to not have a default implementation so that you don't forget to implement it.  (Not that I expect much more new code in Lean 3.)